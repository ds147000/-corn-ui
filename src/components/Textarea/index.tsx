import React from 'react'
import { Textarea as TaroTextarea, View } from '@tarojs/components'
import { TextareaProps as TaroTextareaProps } from '../../types/Textarea'
import { useInput } from '../Input/hook/input'

const CLOSE_MAX_LEN = -1
const DEFAULT_MAX_LEN = 300

export interface TextareaProps extends TaroTextareaProps {
  floor?: React.ReactNode
}

const Textarea: React.FC<TextareaProps> = ({ floor, maxlength = DEFAULT_MAX_LEN, value, onInput, ...props }) => {
  const [ _value, _onInput ] = useInput(value, props.name, onInput)

  return (
    <View className="corn-textarea" >
      <TaroTextarea
        className="corn-textarea-conter"
        value={_value}
        onInput={_onInput}
        maxlength={maxlength}
        {...props}
        data-testid="textarea"
      />
      {maxlength !== CLOSE_MAX_LEN && (
        <View className="corn-textarea-number" >{(_value || '').length}/{maxlength}</View>
      )}
      <View className="corn-textarea-floor" >
        {floor}
      </View>
    </View>
  )
}

export default Textarea
