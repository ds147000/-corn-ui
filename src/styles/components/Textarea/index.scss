.xrk-textarea {
  background: $white;
  border-radius: $pad;
  box-sizing: border-box;
  padding: $pad;

  &-conter {
    display: block;
    width: 100%;
    border: 0;
    outline: 0;
    min-height: $textareaMinHeight;
    font-size: $textareaFontSize;
    font-weight: 400;
    color: $text1;
    line-height: $textareaLineHeight;
    resize: none;
  }

  &-number {
    font-size: 24px;
    font-weight: 400;
    color: $text3;
    line-height: 24px;
    padding-top: $pad;
    text-align: right;
  }
}
