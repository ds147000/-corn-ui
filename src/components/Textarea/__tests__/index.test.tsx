import React from 'react'
import { render, waitFor } from '@testing-library/react'
import UserEvent from '@testing-library/user-event'
import Textarea from '../index'

describe('Textarea', () => {
  test('基础', async () => {
    const screen = render(<Textarea />)
    await waitFor (() => expect(screen.container).toMatchSnapshot())
  })

  test('插入底部节点', async () => {
    const screen = render(<Textarea floor={<div>我是底部</div>} />)
    screen.getByText('我是底部')
    await waitFor (() => expect(screen.container).toMatchSnapshot())
  })

  test('插入底部节点 可操作', async () => {
    const click = jest.fn()
    const screen = render(<Textarea floor={<div onClick={click} >我是底部</div>} />)
    UserEvent.click(screen.getByText('我是底部'))
    await waitFor (() => expect(click).toHaveBeenCalledTimes(1))
  })

  test('输入', async () => {
    const screen = render(<Textarea  />)
    UserEvent.type(screen.getByTestId('textarea'), '我是输入框')
    screen.getByText('我是输入框')
  })

  test('设置字数限制', async () => {
    const screen = render(<Textarea maxlength={6} />)
    UserEvent.type(screen.getByTestId('textarea'), '我是输入框')
    UserEvent.type(screen.getByTestId('textarea'), '123456')
    await waitFor(() => expect((screen.getByTestId('textarea') as HTMLTextAreaElement).value).toBe('我是输入框1'))
  })
})
