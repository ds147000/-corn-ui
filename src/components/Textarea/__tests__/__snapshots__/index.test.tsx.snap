// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea 基础 1`] = `
<div>
  <div
    class="xrk-textarea"
  >
    <textarea
      class="xrk-textarea-conter"
      data-testid="textarea"
      maxlength="300"
      placeholder=""
    />
    <div
      class="xrk-textarea-number"
    >
      0
      /
      300
    </div>
    <div
      class="xrk-textarea-floor"
    />
  </div>
</div>
`;

exports[`Textarea 插入底部节点 1`] = `
<div>
  <div
    class="xrk-textarea"
  >
    <textarea
      class="xrk-textarea-conter"
      data-testid="textarea"
      maxlength="300"
      placeholder=""
    />
    <div
      class="xrk-textarea-number"
    >
      0
      /
      300
    </div>
    <div
      class="xrk-textarea-floor"
    >
      <div>
        我是底部
      </div>
    </div>
  </div>
</div>
`;
